aten/src/ATen/native/Normalization.cpp

```diff
@@ -54,6 +54,60 @@ struct Var {
   }
 };
 
+/// A fast path for CPU inference when all tensors are contiguous.
+template<typename scalar_t>
+void batch_norm_cpu_inference_contiguous(Tensor* output, const Tensor& input,
+    const Tensor& weight /* optional */, const Tensor& bias /* optional */,
+    const Tensor& mean, const Tensor& variance, double eps) {
+  int64_t n_batch = input.size(0);
+  int64_t n_channel = input.size(1);
+  int64_t image_size = input.numel() / n_batch / n_channel;
+
+  scalar_t* output_data = output->data<scalar_t>();
+  const scalar_t* input_data = input.data<scalar_t>();
+  const scalar_t* weight_data = weight.defined() ? weight.data<scalar_t>() : nullptr;
+  const scalar_t* bias_data = bias.defined() ? bias.data<scalar_t>() : nullptr;
+  const scalar_t* mean_data = mean.data<scalar_t>();
+  const scalar_t* var_data = variance.data<scalar_t>();
+
+  /// Collect the linear and constant terms regarding the input.
+  /// output(n, c, h, w)
+  ///     = (input(n, c, h, w) - mean(c)) / sqrt(var(c) + eps) * weight(c)
+  ///         + bias(c)
+  ///     = input(n, c, h, w) * inv_var(c) * weight(c) +
+  ///         - mean(c) * inv_var(c) * weight(c) + bias(c),
+  /// where inv_var(c) = 1 / sqrt(var(c) + eps).
+  /// So the linear term, alpha(c) = inv_var(c) * weight(c),
+  ///   the constant term beta(c) = bias(c) - mean(c) * inv_var(c) * weight(c)
+  Tensor alpha = at::empty_like(mean);
+  Tensor beta = at::empty_like(mean);
+  scalar_t* alpha_data = alpha.data<scalar_t>();
+  scalar_t* beta_data = beta.data<scalar_t>();
+  for (int64_t c = 0; c < n_channel; c++) {
+    scalar_t inv_var = 1 / std::sqrt(var_data[c] + eps);
+    scalar_t weight_v = weight_data ? weight_data[c] : 1;
+    scalar_t bias_v = bias_data ? bias_data[c] : 0;
+    alpha_data[c] = inv_var * weight_v;
+    beta_data[c] = bias_v - mean_data[c] * inv_var * weight_v;
```

Comment:
I don't know how much it matters, but computing the `rsqrt` as a distinct expression is losing precision compared to doing an `fdiv` for each of `alpha_data` and `beta_data`.  It's probably fine if that necessary for performance, but if it's not then it would be nice to preserve the extra precision.

Reply:
This portion is a performance critical. And from other experiments we had before, it shouldn't cause any accuracy problem.

Reply:
This loop is unvectorized in llvm because of the presence of the `sqrt` function (unless `-ffast-math` is set). Benchmark for the image_size=1 case:

```
batch_norm: data shape: [1, 65536, 1], bandwidth: 6.23 GB/s
```

And as a debug I printed out the timing for this loop ("first" loop) versus the loop below ("second" loop):

```
first time 0.000150609
second time 2.1667e-05
```

So, this loop is taking ~6x longer than the loop below. I am working with @ZolotukhinM to see if we can fix this

Reply:
@cpuhrsch we've found that setting `-fno-math-errno` allows the sqrt call to be vectorized. This actually seems fairly safe. Do you think we can turn this flag on in ATen?